### ISSUE TITLE
Differential expression via categorical metadata

### ISSUE THREAD

**Issue by colinmegill** [chanzuckerberg/cellxgene#729]
### Problems:
* Strange to have diffexp / computation in left sidebar with other categorical metadata
* Users cannot return to a selection once they've made it and changed it
* Differential expression selections do not persist between sessions

### Solution:
* Compute differential expression based on selections of categorical metadata
* If the diffexp button (see mock) of one categorical metadata option is selected (ie., Heart), compute cells in this category vs all other cells (invert selection)
* If the diffexp button of two categorical metadata options are selected, compute diffexp comparing the two (ie., Heart vs Lung)
* This issue depends is partly on #524: to perform differential expression which includes continuous or spatial metadata, the user will need to create a new categorical metadata field and use that to compute differential expression. 

While this adds a bit of overhead to computing differential expression for cases involving continuous and spatial information, it allows users to return to those selections as they've been persisted as new categorical metadata fields. It reduces the number of clicks necessary to compute diffexp if only categorical metadata is involved. 

Mock:
![image](https://user-images.githubusercontent.com/1770265/56772647-bcea1900-676f-11e9-92c2-1b95da34fcb8.png)

**Comment by neurojacob** [comment]
Is it possible to make the differential expression analysis tool slightly more intuitive to use? For a layperson, it is not obvious that "1" and "2" will refer to the cells that you lasso or select. Maybe having a 'differential analysis' tab, where that is written out. You could press '1', and it would then tell the user to lasso or select the cells, then repeat for two, and finally calculate differentials. I tried to get several colleagues to use my site and they all had issues figuring it out. It is also unintuitive to me that differentially expressed genes would not be returned as a list, but rather a histogram by default. I would prefer to see the full list rather than the top 10 histograms!

**Comment by ambrosejcarr** [comment]
> If the diffexp button (see mock) of one categorical metadata option is selected (ie., Heart), compute cells in this category vs all other cells (invert selection)

This "one vs all else" test is typically executed to detect "marker genes", or genes that uniquely characterize a particular population. To accomplish this, a more sophisticated test function must be used. See [scanpy documentation](https://scanpy.readthedocs.io/en/stable/api/index.html#marker-genes). In brief, consider the following situation: 

There are populations a, b, c, d, e, and f, each of which has the same number of cells. 
gene 1 is expressed in populations a and b at value "1" and value "0" in the remaining populations. 

A ttest comparing a with (b, c, d, e, f) detect an average expression of 1 in population a and 0.2 in the "other" population. The t-test would report a significant result, however, gene 1 is **not a marker gene** of population a, because it is not uniquely expressed there. 

There are a number of ways to address this problem. Scanpy has some more sophisticated examples, but I give a simple one here to explain how the problem can be solved. A series of tests are run. First, an [ANOVA](https://en.wikipedia.org/wiki/Analysis_of_variance) asks the question "is the expression of gene 1 different in any of these groups?". For genes that pass this test (gene 1 would pass), follow-up pairwise t-tests are run sequentially for population a vs each other population. If and only if a is expressed at significantly heightened expression relative to all other populations, it is returned as a marker gene. In the above example, gene 1 is not expressed at a higher level in population a than population b, so it would be discarded.

**Comment by colinmegill** [comment]
An idea from @aopisco and Lubert Stryer — one could, on the running of differential expression, automatically create a corresponding category with labels named 'pop1 and pop2', where the category name itself is the timestamp that corresponds to the run. cc @seve

**Comment by colinmegill** [comment]
I think that's good enough to close this issue, and it's composable from actions we already have. We already have the creation of two genesets as a 'side effect', so more side effects are not necessarily a liability in the application.

**Comment by colinmegill** [comment]
The disadvantage to this path, is that while fast, it doesn't work well for `lung` vs `heart`

**Comment by colinmegill** [comment]
The solution is likely to have a dropdown the diffexp so that the options are — 

1. do diffexp 
2. do diffexp and create shadow category based on run

**Comment by ambrosejcarr** [comment]
I have consistently had the issue where I circle some cells, assign them to
group 1, circle some other cells, assign them to group 2, compute DE, but
then forget what cells I selected. Am I correct that this suggestion would
address this problem?

If that's the case, I'm wondering how this fits into the longer term vision
of anchoring differential expression to categories. My sense was that the
proposal there was:
1. select cells and create category a
2. select cells and create category b
3. select a and b and select "compute de".

If lung vs heart, then you skip to
3. select a=lung, b=heart, and select "compute de".

**Comment by colinmegill** [comment]
Yes. This proposal addresses the problem (also the cause of this issue in the first place) that users forget what their selection was (quite a large issue). 

* The initial proposal and mock ties diffexp to categorical metadata 
  * it forces users to make a category every time they want to compute custom diffexp that isn't already a category
  * it is very fast in the case that you want 'existing category vs all else' — expand `Tissue` and click diffexp button next to `Lung`
* The amended proposal from Pisco and Stryer suggests that like gene sets, the creation of categories is procedural, and unlike gene sets, it only happens if the user chooses that side effect
  * A simple dropdown on the button would suffice
    * either compute
    * or compute and create category from selection
  * lung vs heart already exists as a category, so just compute
  * lasso region vs lasso region doesn't exist, so compute and create a category with the same params as the gene set name
    * category name: run timestamp
    * label 1: Pop1
    * label 2: Pop2

Quick to implement, since it's composed of existing actions.
